hail/python/hail/methods/statgen.py

```diff
@@ -1343,55 +1333,48 @@ def _logistic_regression_rows_nd(test,
+    # yvecs is a list of sample-length vectors, one for each dependent variable.
+    ht = ht.annotate_globals(yvecs=[hl.nd.array(ht.samples[y_name]) for y_name in y_field_names])
 
     # Fit null models, which means doing a logreg fit with just the covariates for each phenotype.
-    null_models = hl.range(num_y_fields).map(
-        lambda idx: logreg_fit(ht.cov_nd, ht.y_nd[:, idx], None, max_iter=max_iterations, tol=tolerance))
-    ht = ht.annotate_globals(nulls=null_models)
-    ht = ht.transmute(x=hl.nd.array(mean_impute(ht.entries[x_field_name])))
+    def fit_null(yvec):
+        null_fit = logreg_fit(ht.covmat, yvec, None, max_iter=max_iterations, tol=tolerance)
+        return (
+            hl.case()
+            .when(~null_fit.exploded,
+                  (hl.case()
+                   .when(null_fit.converged, null_fit)
+                   .or_error("Failed to fit logistic regression null model (standard MLE with covariates only): "
+                             "Newton iteration failed to converge")))
+            .or_error(hl.format("Failed to fit logistic regression null model (standard MLE with covariates only): "
+                                "exploded at Newton iteration %d", null_fit.num_iter)))
+    ht = ht.annotate_globals(null_fits=ht.yvecs.map(fit_null))
 
-    if test == "wald":
-        test_func = functools.partial(wald_test, max_iter=max_iterations, tol=tolerance)
-    elif test == "lrt":
-        test_func = functools.partial(lrt_test, max_iter=max_iterations, tol=tolerance)
-    elif test == "score":
-        test_func = logistic_score_test
-    elif test == "firth":
-        test_func = functools.partial(firth_test, max_iter=max_iterations, tol=tolerance)
-    else:
-        raise ValueError(f"Illegal test type {test}")
-
-    def test_against_null(covs_and_x, y_vec, null_fit, name):
-        return (hl.case()
-                .when(~null_fit.exploded,
-                      (hl.case()
-                       .when(null_fit.converged, test_func(covs_and_x, y_vec, null_fit, name))
-                       .or_error("Failed to fit logistic regression null model (standard MLE with covariates only): Newton iteration failed to converge")))
-                .or_error(hl.format("Failed to fit logistic regression null model (standard MLE with covariates only): exploded at Newton iteration %d", null_fit.num_iter)))
-
-    covs_and_x = hl.nd.hstack([ht.cov_nd, ht.x.reshape((-1, 1))])
-    test_structs = hl.range(num_y_fields).map(
-        lambda idx: test_against_null(covs_and_x, ht.y_nd[:, idx], ht.nulls[idx], "logistic")
-    )
-    ht = ht.annotate(logistic_regression=test_structs)
+    ht = ht.transmute(x=hl.nd.array(mean_impute(ht.entries[x_field_name])))
+    covs_and_x = hl.nd.hstack([ht.covmat, ht.x.reshape((-1, 1))])
+
+    def test(yvec, null_fit):
+        if test == 'score':
+            return logistic_score_test(covs_and_x, y, null_fit)
```

Comment:
Should this be `yvec`?

Reply:
eagle eyes, yes it absolutely should be. Luckily this was caught by the tests as a mismatched source objects error.